processor/src/operations/sys_ops.rs

```diff
@@ -122,9 +131,62 @@ impl Process {
     {
         self.stack.copy_state(0);
         self.decoder.set_user_op_helpers(Operation::Emit(event_id), &[event_id.into()]);
-        host.on_event(self.into(), event_id)?;
 
-        Ok(())
+        // If it's a native event, handle it directly. Otherwise, forward it to the host.
+        if let Some(advice_injector) = AdviceInjector::from_event_id(event_id) {
+            self.handle_advice_injector(advice_injector, host)
+        } else {
+            host.on_event(self.into(), event_id)
+        }
+    }
+
+    fn handle_advice_injector(
+        &self,
+        advice_injector: AdviceInjector,
+        host: &mut impl Host,
+    ) -> Result<(), ExecutionError> {
+        let advice_provider = host.advice_provider_mut();
+        let process_state: ProcessState = self.into();
+        match advice_injector {
+            AdviceInjector::MerkleNodeMerge => merge_merkle_nodes(advice_provider, process_state),
+            AdviceInjector::MerkleNodeToStack => {
+                copy_merkle_node_to_adv_stack(advice_provider, process_state)
+            },
+            AdviceInjector::MapValueToStack => {
+                copy_map_value_to_adv_stack(advice_provider, process_state, false)
+            },
+            AdviceInjector::MapValueToStackN => {
+                copy_map_value_to_adv_stack(advice_provider, process_state, true)
+            },
+            AdviceInjector::UpdateMerkleNode => {
+                let _ = update_operand_stack_merkle_node(advice_provider, process_state)?;
+                Ok(())
+            },
```

Comment:
I'm wondering if we actually need the `AdviceInjector::UpdateMerkleNode`. I think `mrupdate` instruction already provides this functionality. Maybe we should get rid of it?

Reply:
Agreed, removed